README.md
# extract-authentication-middleware

# [![extract-authentication-middleware](docs/carbon.png)](https://git.nicolaischmid.de/wasc-npm/extract-authentication-middleware)

## Overview

[extract-authentication-middleware](https://git.nicolaischmid.de/wasc-npm/extract-authentication-middleware) is an express middleware for getting the authorization header of a request and its embeded JWT Token and decoding its content.

## Usage

Just like above:

```javascript
import checkAuthentication from '@private/extract-authentication-middleware'
import express from 'express';

const app = express();

app.use(checkAuthentication({
  secret: process.env.AUTHENTICATION_SECRET,
  error: pino.error,
});
```

Import the module, pass it the secret for verifying the token and use it. It works with CJS `require` as well.

## Configuration

- `secret` The secret for verifying the tokens signature
- `error` A error logging function for logging any non expected errors [default: pino.error]

## Data

If the authentication is successfull the middleware will create a few properties on the requets object:

```javascript
// request
{
    // ...
    scope: String, // The scope of the authication token
    user: {}, // The encoded user object
    auth: {
        isAuthenticated: true // false
        scope: String, // The scope like above
    }
    // ...
}
```

If the authentication fails, because the token is faulty, not signing properly or if there's any other error, the middleware terminates the request and returns either one of the following three responses:

1.  If the token was used past its expiration:

```javascript
// Status 401
{
    name: 'TokenExpired',
    message: 'The token cannot be used past its expiration',
}
```

2.  If the decoding of the token resulted in any errors, like untrusted signatures, malformed tokens or by providing an empty string as the token:

```javascript
// Status 401
{
    name: 'BadCredentials',
    message: 'The provided token could not be verified.',
}
```

3.  And if any other error occured, it will be logged using the provided logger or pino and the response is the following:

```javascript
// Status 400
{
    name: 'BadRequest',
    message: 'BadRequest',
}
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

<a name="1.0.0"></a>

# [1.0.0](http://git.nicolaischmid.de/wasc-npm/check-auth-header/compare/v0.5.2...v1.0.0) (2019-06-21)

### Features

- implement jwk based verification ([726aa4e](http://git.nicolaischmid.de/wasc-npm/check-auth-header/commits/726aa4e))

<a name="0.5.2"></a>

## [0.5.2](http://git.nicolaischmid.de/wasc-npm/check-auth-header/compare/v0.5.1...v0.5.2) (2018-08-29)

### Bug Fixes

- removed authenticated:false field, to work better together with other auth middlewares ([47ccd47](http://git.nicolaischmid.de/wasc-npm/check-auth-header/commits/47ccd47))

<a name="0.5.1"></a>

## [0.5.1](http://git.nicolaischmid.de/wasc-npm/check-auth-header/compare/v0.5.0...v0.5.1) (2018-07-18)

### Bug Fixes

- added auth property to request even if it is not authenticated ([ba39a46](http://git.nicolaischmid.de/wasc-npm/check-auth-header/commits/ba39a46))

<a name="0.5.0"></a>

# [0.5.0](http://git.nicolaischmid.de/wasc-npm/check-auth-header/compare/v0.4.2...v0.5.0) (2018-07-18)

### Features

- added new properties to request object to work with graphql-auth package ([33c6595](http://git.nicolaischmid.de/wasc-npm/check-auth-header/commits/33c6595))

<a name="0.4.2"></a>

## [0.4.2](http://git.nicolaischmid.de/wasc-npm/check-auth-header/compare/v0.4.1...v0.4.2) (2018-04-06)

<a name="0.4.1"></a>

## [0.4.1](http://git.nicolaischmid.de/wasc-npm/check-auth-header/compare/v0.4.0...v0.4.1) (2018-04-04)

### Bug Fixes

- fixed request.get after auto correction by intellisense ([df48a9a](http://git.nicolaischmid.de/wasc-npm/check-auth-header/commits/df48a9a))

<a name="0.4.0"></a>

# [0.4.0](http://git.nicolaischmid.de/wasc-npm/check-auth-header/compare/v0.3.0...v0.4.0) (2018-04-02)

### Features

- updated name to extract-authentication-middleware ([fc35bee](http://git.nicolaischmid.de/wasc-npm/check-auth-header/commits/fc35bee))


src/index.js
import cookie from 'cookie';
import errs from 'errs';
import getPino from 'pino';

import syncKey from './syncKey';
import asyncKey from './asyncKey';

const pino = getPino();

/**
 * The middleware function. If both secret and jwk are defined jwk will take precedent
 * @param {Object} options
 * @param {String} secret The JWT verification secret
 * @param {String|Object} jwk For JWK based authentication an endpoint or a direct object
 * @param {String} extractName The name of the header or cookie to extract the token from (default: authorization)
 * @param {Object}
 * @param {Function} error The error logging function
 * @returns {Promise[Function]} The actual middleware function
 */
export default async function (options = {}) {
  let logger = pino.error;
  if (!options.secret && !options.jwk) {
    throw errs.create({
      name: 'NoVerfifaction',
      message:
        'No secret or jwk provided, the JWT token cannot be verified properly',
    });
  }

  if (options.error) {
    logger = options.error;
  }

  let verify;

  if (options.secret) verify = syncKey(options.secret);

  if (options.jwk) verify = await asyncKey(options.jwk);

  /**
   * The authentication middleware
   * @param {Request} request The express Request Object
   * @param {Response} response The express Response Object
   * @param {Function} next The express next function
   * @return
   */
  return async function (request, response, next) {
    const extractName = options.extractName || 'authorization';
    let token = request.get(extractName);
    if (!token) {
      const cookies = request.headers.cookie;

      if (cookies) {
        token = cookie.parse(cookies)[extractName];
      }
    }

    try {
      if (!token) return next();

      const decoded = await verify(token);

      request.scope = decoded.scope;
      request.user = decoded.user;

      request.auth = {
        isAuthenticated: true,
        scope: decoded.scope,
      };

      return next();
    } catch (error) {
      if (error.message === 'no key found') {
        return response.status(401).json({
          name: 'BadCredentials',
          message: 'The provided token could not be verified.',
        });
      }
      if (error.name === 'TokenExpiredError') {
        return response.status(401).json({
          name: 'TokenExpired',
          message: 'The token cannot be used past its expiration',
        });
      }
      if (error.name === 'JsonWebTokenError') {
        return response.status(401).json({
          name: 'BadCredentials',
          message: 'The provided token could not be verified.',
        });
      }

      logger(error);

      return response.status(400).json({
        name: 'BadRequest',
        message: 'BadRequest',
      });
    }
  };
}
